File: api/Enkl.Api/Controllers/RetrospectivesController.cs
using Enkl.Api.Dtos;
using Enkl.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Enkl.Api.Controllers;

[ApiController]
[Authorize(Policy = "ProjectMember")]
[Route("api/projects/{projectId:guid}/retrospectives")]
public class RetrospectivesController : ControllerBase
{
    private readonly RetrospectiveService _retrospectives;

    public RetrospectivesController(RetrospectiveService retrospectives)
    {
        _retrospectives = retrospectives;
    }

    [HttpPost]
    public async Task<IActionResult> Create(Guid projectId, CreateRetrospectiveRequest request)
    {
        var result = await _retrospectives.CreateAsync(projectId, request);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpPut("{retrospectiveId:guid}")]
    public async Task<IActionResult> Update(Guid projectId, Guid retrospectiveId, UpdateRetrospectiveRequest request)
    {
        var result = await _retrospectives.UpdateAsync(projectId, retrospectiveId, request);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpDelete("{retrospectiveId:guid}")]
    public async Task<IActionResult> Delete(Guid projectId, Guid retrospectiveId)
    {
        return await _retrospectives.DeleteAsync(projectId, retrospectiveId) ? NoContent() : NotFound();
    }

    [HttpPost("{retrospectiveId:guid}/items")]
    public async Task<IActionResult> CreateItem(Guid projectId, Guid retrospectiveId, CreateRetrospectiveItemRequest request)
    {
        var result = await _retrospectives.CreateItemAsync(projectId, retrospectiveId, request);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpPut("{retrospectiveId:guid}/items/{itemId:guid}")]
    public async Task<IActionResult> UpdateItem(Guid projectId, Guid retrospectiveId, Guid itemId, UpdateRetrospectiveItemRequest request)
    {
        var result = await _retrospectives.UpdateItemAsync(projectId, retrospectiveId, itemId, request);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpDelete("{retrospectiveId:guid}/items/{itemId:guid}")]
    public async Task<IActionResult> DeleteItem(Guid projectId, Guid retrospectiveId, Guid itemId)
    {
        return await _retrospectives.DeleteItemAsync(projectId, retrospectiveId, itemId) ? NoContent() : NotFound();
    }

    [HttpPost("{retrospectiveId:guid}/items/{itemId:guid}/promote")]
    public async Task<IActionResult> PromoteItem(Guid projectId, Guid retrospectiveId, Guid itemId, PromoteRetrospectiveItemRequest request)
    {
        var result = await _retrospectives.PromoteItemAsync(projectId, retrospectiveId, itemId, request);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpPost("{retrospectiveId:guid}/action-items")]
    public async Task<IActionResult> CreateActionItem(Guid projectId, Guid retrospectiveId, CreateRetrospectiveActionItemRequest request)
    {
        var result = await _retrospectives.CreateActionItemAsync(projectId, retrospectiveId, request);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpPut("{retrospectiveId:guid}/action-items/{itemId:guid}")]
    public async Task<IActionResult> UpdateActionItem(Guid projectId, Guid retrospectiveId, Guid itemId, UpdateRetrospectiveActionItemRequest request)
    {
        var result = await _retrospectives.UpdateActionItemAsync(projectId, retrospectiveId, itemId, request);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpDelete("{retrospectiveId:guid}/action-items/{itemId:guid}")]
    public async Task<IActionResult> DeleteActionItem(Guid projectId, Guid retrospectiveId, Guid itemId)
    {
        return await _retrospectives.DeleteActionItemAsync(projectId, retrospectiveId, itemId) ? NoContent() : NotFound();
    }
}

File: php-api/src/Controllers/RetrospectivesController.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Controllers;

use Enkl\Api\Db\Database;
use Enkl\Api\Services\PrincipleService;
use Enkl\Api\Services\RetrospectiveService;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;

/** Ported from Controllers/RetrospectivesController.cs. */
final class RetrospectivesController extends BaseController
{
    private function service(): RetrospectiveService
    {
        $db = Database::connection();
        return new RetrospectiveService($db, new PrincipleService($db));
    }

    public function create(Request $request, Response $response, array $args): Response
    {
        $result = $this->service()->create($args['projectId'], $this->body($request));
        return $result === null ? $this->notFound($response) : $this->json($response, $result);
    }

    public function update(Request $request, Response $response, array $args): Response
    {
        $result = $this->service()->update($args['projectId'], $args['id'], $this->body($request));
        return $result === null ? $this->notFound($response) : $this->json($response, $result);
    }

    public function delete(Request $request, Response $response, array $args): Response
    {
        return $this->service()->delete($args['projectId'], $args['id']) ? $this->noContent($response) : $this->notFound($response);
    }

    public function createItem(Request $request, Response $response, array $args): Response
    {
        $result = $this->service()->createItem($args['projectId'], $args['id'], $this->body($request));
        return $result === null ? $this->notFound($response) : $this->json($response, $result);
    }

    public function updateItem(Request $request, Response $response, array $args): Response
    {
        $result = $this->service()->updateItem($args['projectId'], $args['id'], $args['itemId'], $this->body($request));
        return $result === null ? $this->notFound($response) : $this->json($response, $result);
    }

    public function deleteItem(Request $request, Response $response, array $args): Response
    {
        return $this->service()->deleteItem($args['projectId'], $args['id'], $args['itemId']) ? $this->noContent($response) : $this->notFound($response);
    }

    public function promoteItem(Request $request, Response $response, array $args): Response
    {
        $result = $this->service()->promoteItem($args['projectId'], $args['id'], $args['itemId'], $this->body($request));
        return $result === null ? $this->notFound($response) : $this->json($response, $result);
    }

    public function createActionItem(Request $request, Response $response, array $args): Response
    {
        $result = $this->service()->createActionItem($args['projectId'], $args['id'], $this->body($request));
        return $result === null ? $this->notFound($response) : $this->json($response, $result);
    }

    public function updateActionItem(Request $request, Response $response, array $args): Response
    {
        $result = $this->service()->updateActionItem($args['projectId'], $args['id'], $args['itemId'], $this->body($request));
        return $result === null ? $this->notFound($response) : $this->json($response, $result);
    }

    public function deleteActionItem(Request $request, Response $response, array $args): Response
    {
        return $this->service()->deleteActionItem($args['projectId'], $args['id'], $args['itemId']) ? $this->noContent($response) : $this->notFound($response);
    }
}
